src/replica/disk_migration/replica_disk_migrator.cpp

```diff
@@ -0,0 +1,151 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "replica/replica_stub.h"
+#include "replica_disk_migrator.h"
+
+#include <dsn/dist/fmt_logging.h>
+
+namespace dsn {
+namespace replication {
+
+replica_disk_migrator::replica_disk_migrator(replica *r) : replica_base(r), _replica(r) {}
+
+replica_disk_migrator::~replica_disk_migrator() = default;
+
+// THREAD_POOL_REPLICATION
+void replica_disk_migrator::on_migrate_replica(const replica_disk_migrate_request &req,
+                                               /*out*/ replica_disk_migrate_response &resp)
+{
+    if (!check_migration_args(req, resp)) {
+        return;
+    }
+
+    _status = disk_migration_status::MOVING;
+    ddebug_replica("received replica disk migrate request({}), update status from {}=>{}",
+                   _request_msg,
+                   enum_to_string(disk_migration_status::IDLE),
+                   enum_to_string(status()));
+
+    tasking::enqueue(
+        LPC_REPLICATION_LONG_COMMON, _replica->tracker(), [=]() { migrate_replica(req); });
+}
+
+bool replica_disk_migrator::check_migration_args(const replica_disk_migrate_request &req,
+                                                 /*out*/ replica_disk_migrate_response &resp)
+{
+    _replica->_checker.only_one_thread_access();
+
+    _request_msg = fmt::format(
+        "gpid={}, origin={}, target={}", req.pid.to_string(), req.origin_disk, req.target_disk);
+
+    // TODO(jiashuo1) may need manager control migration flow
+    if (status() != disk_migration_status::IDLE) {
+        std::string err_msg =
+            fmt::format("Existed migrate task({}) is running", enum_to_string(status()));
+        derror_replica(
+            "received replica disk migrate request({}), err = {}", _request_msg, err_msg);
+        resp.err = ERR_BUSY;
+        resp.hint = err_msg;
+        return false;
+    }
+
+    if (_replica->status() != partition_status::type::PS_SECONDARY) {
+        std::string err_msg =
+            fmt::format("Invalid partition status({})", enum_to_string(_replica->status()));
+        derror_replica(
+            "received replica disk migrate request({}), err = {}", _request_msg, err_msg);
+        resp.err = ERR_INVALID_STATE;
+        resp.hint = err_msg;
+        return false;
+    }
+
+    if (req.origin_disk == req.target_disk) {
+        std::string err_msg = fmt::format(
+            "Invalid disk tag(origin({}) equal target({}))", req.origin_disk, req.target_disk);
+        derror_replica(
+            "received replica disk migrate request({}), err = {}", _request_msg, err_msg);
+        resp.err = ERR_INVALID_PARAMETERS;
+        resp.hint = err_msg;
+        return false;
+    }
+
+    bool valid_origin_disk = false;
+    bool valid_target_disk = false;
+    for (const auto &dir_node : _replica->get_replica_stub()->_fs_manager._dir_nodes) {
```

Comment:
I recommend that adding following comment here:
`// std::vector<std::shared_ptr<dir_node>>` to show the structure of `_replica->get_replica_stub()->_fs_manager._dir_nodes`